pkg/dns/dns.go

```diff
@@ -0,0 +1,310 @@
+		}
+	}
+
+	return r, nil
+}
+
+func (r *DNSResolver) StartDNSResolver(stopCh <-chan struct{}) {
+	go r.startResolver()
+	go r.refreshDNS()
+	go func() {
+		<-stopCh
+		r.dnsRefreshQueue.ShutDown()
+		close(r.DnsResolverChan)
+	}()
+}
+
+// startResolver watches the DnsResolver Channel
+func (r *DNSResolver) startResolver() {
+	temp := make(chan struct{}, MaxConcurrency)
+	var wg sync.WaitGroup
+	for domains := range r.DnsResolverChan {
+		temp <- struct{}{}
+		wg.Add(1)
+
+		go func(domains map[string]time.Duration) {
+			defer wg.Done()
+			defer func() { <-temp }()
+			r.resolveDomains(domains)
+		}(domains)
+	}
+	wg.Wait()
+}
+
+// resolveDomains takes a map of hostnames and refresh rate
+func (r *DNSResolver) resolveDomains(domains map[string]time.Duration) {
+	// Stow domain updates, need to remove unwatched domains first
+	r.removeUnwatchedDomain(domains)
+
+	for name, refreshRate := range domains {
+		addrs := r.resolve(name, refreshRate)
+		log.Debugf("resolve dns , name: %s, addr: %v\n", name, addrs)
+		r.updateBPFMap(name, addrs)
+	}
+}
+
+// removeUnwatchedDomain cancels any scheduled re-resolve for names we no longer care about
+func (r *DNSResolver) removeUnwatchedDomain(domains map[string]time.Duration) {
+	r.Lock()
+	defer r.Unlock()
+	for name := range r.cache {
+		if _, ok := domains[name]; ok {
+			continue
+		}
+		delete(r.cache, name)
+		r.deleteBPFMap(name)
+	}
+}
+
+// This functions were copied and adapted from github.com/istio/istio/pilot/pkg/model/network.go.
+func (r *DNSResolver) resolve(name string, refreshRate time.Duration) []string {
+	if entry, ok := r.cache[name]; ok && entry.expiry.After(time.Now()) {
+		return entry.value
+	}
+
+	r.Lock()
+	defer r.Unlock()
+	// ideally this will not happen more than once for each name and the cache auto-updates in the background
+	// even if it does, this happens on the SotW ingestion path (kube or meshnetworks changes)
+	entry := r.cache[name]
+	delete(r.cache, name)
+	addrs, ttl, err := r.doResolve(name, refreshRate)
+	// refresh the dns address periodically by respecting the dnsRefreshRate and ttl, which one is shorter
+	if ttl > refreshRate {
+		ttl = refreshRate
+	}
+	expiry := time.Now().Add(ttl)
+	if err != nil {
+		// gracefully retain old addresses in case the DNS server is unavailable
+		addrs = entry.value
+	}
+
+	r.cache[name] = domainCacheEntry{
+		value:  addrs,
+		expiry: expiry,
+	}
+
+	dr := &domainWithRefreshRate{
+		name:        name,
+		refreshRate: refreshRate,
+	}
+	// push torefresh queue one second before dns expire
+	r.dnsRefreshQueue.AddAfter(dr, time.Until(expiry))
+	return addrs
+}
+
+// refreshDNS is triggered via time.AfterFunc and will recursively schedule itself that way until timer is cleaned
+// up via removeUnwatchedDomain.
+func (r *DNSResolver) refreshDNS() {
+	for {
+		element, quit := r.dnsRefreshQueue.Get()
+		if quit {
+			return
+		}
+		r.RLock()
+		dr := element.(*domainWithRefreshRate)
+		old := r.cache[dr.name]
```

Comment:
This can be nil, because of removeUnwatchedxxx.

So if this domain not exist, stop resolve it, and not need to re enqueue either